linker-interface/shared/src/main/scala/org/scalajs/linker/interface/StandardConfig.scala

```diff
@@ -158,6 +158,22 @@ final class StandardConfig private (
 object StandardConfig {
   import StandardConfigPlatformExtensions.ConfigExt
 
+  def fingerprint(config: StandardConfig): String =
+    s"""StandardConfig(
+       |semantics=${Semantics.fingerprint(config.semantics)},
+       |moduleKind=${ModuleKind.fingerprint(config.moduleKind)},
+       |esFeatures=${ESFeatures.fingerprint(config.esFeatures)},
+       |checkIR=${config.checkIR},
+       |optimizer=${config.optimizer},
+       |parallel=${config.parallel},
+       |sourceMap=${config.sourceMap},
+       |relativizeSourceMapBase=
+       |${config.relativizeSourceMapBase.map(_.toASCIIString())},
+       |closureCompilerIfAvailable=${config.closureCompilerIfAvailable},
+       |prettyPrint=${config.prettyPrint},
+       |batchMode=${config.batchMode}
+       |)""".stripMargin.replaceAll("\n", "")
+
```

Comment:
I'd like us to explore other solutions to generate fingerprints. The main problem I have with the current approach is that it is difficult to see if the strings in a fingerprint are properly derived. For example, is `config.checkIR` ok? I need to go and look at the type of `checkIR` and then at its `toString` method to know. That's not ideal.
There is of course also the issue that one has to manually check all fields are there, but I don't think this can be avoided without macros (which we want to avoid more).

Did you have a look at the serializer API in test-common? If yes, is it infeasible to use it?

Reply:
> Did you have a look at the serializer API in test-common? If yes, is it infeasible to use it?

The linker interface should definitely not depend on test-common. ;)

Reply:
Well yes, but we can move the code to somewhere else or source include it. It's just an implementation detail here, isn't it?

Reply:
That looks like overengineering to me. Sourcing-in is not acceptable here because they would end up twice on classpaths that depend both on the linker interface and the test adapter (including our own sbt plugin!). And creating yet another artifact for this looks overkill, when simply using `toString`s and `fingerprint`s is basically sufficient.

Yes, there exists a possibility that we're going to use an implicit `toString()` instead of a `fingerprint()` by mistake. Is preventing ourselves from that mistake worth so much infrastructure? I don't think so.

Reply:
OK. Fair enough. But can we at least find a solution that makes it reasonably easy to visually inspect that a `fingerprint` is correct? With the current code, it is very hard and that, combined with it being hard to test, will set us up for likely bugs.

Reply:
What do you think of the following approach?

We could add a `FingerprintBuilder` with overloaded methods for each "fingerprintable" type:

```scala
class FingerprintBuilder(name: String) {
  def addField(name: String, value: Boolean): FingerprintBuilder = ???
  def addField(name: String, value: String): FingerprintBuilder = ???
  def build(): String = ???
}
```

This would then be called as follows:

```scala
object StandardConfig {
  def fingerprint(config: StandardConfig): String = {
    new FingerprintBuilder("StandardConfig")
      .addField("semantics", Semantics.fingerprint(config.semantics))
      // ...
      .addField("batchMode", config.batchMode)
      .build()
  }
}
```

There are perhaps more fancy variants of this idea, with `Fingerprint` wrappers or tagged types, or `Fingerprint[T]` typeclasses and the like, but I think the above might be good enough. 

The `FingerprintBuilder` could be tested with Scalacheck or something like that. Centralizing the fingerprinting logic also allows us to more easily change the type of a fingerprint `Array[Byte]` or something else if we choose to.

Reply:
Oh, I like this approach. Will prevent most of the mishaps (implicit to string, reasonably OK to inspect for missing fields). @sjrd WDYT?

Just one thing, your comment says "with overloaded methods for each "fingerprintable" type:". However, the code suggests that a `FingerprintBuilder` only accepts primitives (hence the explicit call to `Semantics.fingerprint`). I think how the code currently is would be better. If `FingerprintBuilder` knows about all the fingerprintable types, that'd be kind of strange.

Reply:
Oh, that's what I meant, but I see that how I wrote it isn't super clear. Indeed, the `FingerprintBuilder` shouldn't know about the linker types, but should only accept primitives. If needed, the implementations of `fingerprint` should explicitly call other `fingerprint` methods to convert members to primitives that `FingerprintBuilder` can use.

Reply:
I spoke with @sjrd and he said he preferred an approach based on typeclasses, which is what I've implemented in 328801eabde719e7ac1b57593d3b541ec04612fa.

The problem with the approach I had previously suggested is that it exposes data in an intermediary form (string or boolean), which he said felt less clean than encapsulating the transformation to fingerprints in a typeclass (let me know if I've oversimplified the argument, @sjrd).

This new approach gets rid of all `private[interface] def fingerprint`, except for one in `RuntimeClassNameMapper`, which is needed to expose the private subtypes of `RuntimeClassNameMapper` as fingerprints.

Reply:
That is of course an even more luxurious approach. I'm absolutely OK with that. Some less-higher level comments:

- Have you considered moving `Fingerprint` to `unstable` (but still make it `private[interface]`)? (I'm genuinely not sure what the right place is her).
- It feels strange to me to separate the fingerprinting of an individual class from it's definition. Have you considered defining the typeclasses (except for primitive types) in the companions of the respective classes?
- Have you considered using `implicit object` instead of `implicit val` + anonymous class?

Reply:
- I don't think `Fingerprint` in its current form needs to go in `unstable`, as it's `private[interface]` anyway. What could however be marked "unstable" are the public parts of fingerprinting, namely the two public `fingerprint` methods (on `StandardConfig` and `ModuleInitializer`). I could mark those as unstable by:
  1. Moving `Fingerprint` to `unstable`
  2. Making `Fingerprint` public and making its members `private[interface]`
  3. Moving the two public  fingerprinting methods (on `StandardConfig` and `ModuleInitializer`) to `Fingerprint` and making them public.

  What do you think of that approach?
- I'm okay with that. Changed in 2e52230891c0352c29325346112ae1fd89e911e0
- That's a good idea, it's terser and probably a little simpler. Changed in 2e52230891c0352c29325346112ae1fd89e911e0

Reply:
> What do you think of that approach?

Ah, no the `fingerprint` higher level method needs to be in `interface`. It's a stable API we want to commit to.

So this approach is good with me :) I'll make some more detailed comments.